src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics.Hashing;
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
+    /// Provides an immutable list implementation which implements sequence equality.
+    /// </summary>
+    public sealed class ImmutableEquatableArray<T> : IEquatable<ImmutableEquatableArray<T>>, IReadOnlyList<T>
+        where T : IEquatable<T>
+    {
+        public static ImmutableEquatableArray<T> Empty { get; } = new ImmutableEquatableArray<T>(Array.Empty<T>());
+
+        private readonly T[] _values;
+        public T this[int index] => _values[index];
+        public int Count => _values.Length;
+
+        public ImmutableEquatableArray(IEnumerable<T> values)
+            => _values = values.ToArray();
+
+        public bool Equals(ImmutableEquatableArray<T>? other)
+            => other != null && ((ReadOnlySpan<T>)_values).SequenceEqual(other._values);
+
+        public override bool Equals(object? obj)
+            => obj is ImmutableEquatableArray<T> other && Equals(other);
+
+        public override int GetHashCode()
+        {
+            int hash = 0;
+            foreach (T value in _values)
+            {
+                hash = HashHelpers.Combine(hash, value is null ? 0 : value.GetHashCode());
+            }
+
+            return hash;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(_values);
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => ((IEnumerable<T>)_values).GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable<T>)_values).GetEnumerator();
+
+        public struct Enumerator
+        {
+            private readonly T[] _values;
+            private int _index;
+
+            internal Enumerator(T[] values)
+            {
+                _values = values;
+                _index = -1;
+            }
+
+            public bool MoveNext() => ++_index < _values.Length;
```

Comment:
We wouldn't do this if this were a publicly consumable type (call MoveNext enough after it returns false and it could start returning true again).  But presumably even though this is public the intent here is that no one is using it other than us?

Reply:
I could try improving this a bit, since most likely we'd need to reuse that implementation for other source generators as well.